src/lua/protobuf.lua

```diff
@@ -657,9 +726,142 @@ end
+            local value = decode_field(protocol, field_def, ctx)
+            table.insert(res, value)
+            if ctx.pos > string.len(ctx.val) then
+                return res
+            else
+                local _, res_id = wireformat.decode_tag(ctx)
+                field_id = res_id
+            end
+        until field_id ~= field_def.id
+        ctx.pos = ctx.pos - 1
+    end
+    return res
+end
+
+decode_field = function(protocol, field_def, ctx)
+    local res
+    if field_def.repeated and not ctx.ignore_repeated then
+        res = decode_repeated(protocol, field_def, ctx)
+    elseif is_scalar(field_def) then
+        local scalar_def = scalars[field_def.type]
+        res = scalar_def.decode(ctx)
+        validate_scalar(field_def, res)
+    elseif is_enum(protocol, field_def) then
+        local id = wireformat.decode_int(ctx)
+        local curr_enum = protocol[field_def.type]
+        if curr_enum.value_by_id[id] then
+            res = curr_enum.value_by_id[id]
+        else
+            res = id
+        end
+    elseif is_message(protocol, field_def) then
+        local msg_data = wireformat.decode_len(ctx)
+        res = decode(protocol, field_def.type, msg_data, ctx.opts)
+    end
+    return res
+end
+
+decode = function(protocol, message_name, data, opts)
+    local opts = opts or {}
+    assert(type(opts) == 'table')
+    local res = {}
+    local ctx = {val = data, pos = 1, opts = opts}
+    local message_def = protocol[message_name]
+    if message_def == nil then
+        error(('There is no message or enum named %q in the given protocol')
+            :format(message_name))
+    end
+    if message_def.type ~= 'message' then
+        assert(message_def.type == 'enum')
+        error(('Attempt to decode enum %q as a top level message'):format(
+            message_name))
+    end
+    local field_by_id = message_def.field_by_id
+    while ctx.pos <= string.len(ctx.val) do
+        -- Read tag from input get type and field_id
+        local type, field_id = wireformat.decode_tag(ctx)
+        local field_def = field_by_id[field_id]
+        validate_field_id(field_id)
+        validate_decoded_type(protocol, field_def, type)
+        if field_def == nil then
+            local unknown_field = read_unknown_field(ctx, type)
+            if res['_unknown_fields'] == nil then
+                res['_unknown_fields'] = {}
+            end
+            table.insert(res['_unknown_fields'], unknown_field)
+        else
+        local decode_from_packed = false
+        if field_def.repeated and is_scalar(field_def) then
+                if type == WIRE_TYPE_LEN and field_def.type ~= 'string'
+                    and field_def.type ~= 'bytes' then
+                        decode_from_packed = true
+                end
+            end
+        ctx.ignore_repeated = false
+        ctx.packed = decode_from_packed
+            res[field_def.name] = decode_field(protocol, field_def, ctx)
+        end
```

Comment:
Something is definitely bad with indent here.

Reply:
Thanks, fixed indentation here.